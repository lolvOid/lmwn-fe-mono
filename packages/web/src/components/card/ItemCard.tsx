import { getThaiBaht } from '@/utils/formatCurrency';
import { getDiscountedPrice } from '@/utils/helper';
import { SyntheticEvent } from 'react';
import { FaPlus } from 'react-icons/fa';
interface ItemCardProps {
    imageSource: string;
    name: string;
    fullPrice: number;
    discountedPercent: number;
    totalInStock: number;
    onClick?: (event: SyntheticEvent) => void;
}

const ItemCard = ({
    name,
    imageSource,
    fullPrice,
    discountedPercent,
    totalInStock,
    onClick,
}: ItemCardProps) => {
    return (
        <div
            data-testid="item-card"
            role="button"
            className={`relative flex lg:flex-row flex-col lg:justify-between border-solid border-gray-200 border-[0.4px] p-2 rounded-md ${totalInStock != 0 || totalInStock != undefined ? 'opacity-100 cursor-pointer hover:bg-gray-100' : 'opacity-50 cursor-none'}`}
            onClick={onClick}
            onKeyDown={() => {}}
            tabIndex={0}
        >
            <div className="lg:w-32 w-full max-w-40 h-32 flex-shrink-0 justify-center items-center">
                <img
                    src={
                        imageSource ||
                        'https://dummyimage.com/128x128/dedede/5e5e5e.jpg&text=No+Image+Available'
                    }
                    alt={name || 'Item'}
                    className="w-full h-full object-cover object-center"
                />
            </div>
            <div className="px-2 py-2 flex lg:flex-col w-full flex-grow lg:items-end items-start">
                <div className="w-full flex flex-col text-sm flex-grow gap-1">
                    <span className="text-xs text-wrap" data-testid="item-card-name">
                        {name || 'Item'}
                    </span>
                    <span
                        className="text-sm primary-text text-wrap"
                        data-testid="item-card-discounted-price"
                    >
                        {discountedPercent
                            ? getThaiBaht(getDiscountedPrice(fullPrice, discountedPercent))
                            : getThaiBaht(fullPrice)}
                    </span>
                    {discountedPercent !== 0 && (
                        <span
                            className="text-xs text-wrap text-gray-600 line-through"
                            data-testid="item-card-price"
                        >
                            {discountedPercent && getThaiBaht(fullPrice)}
                        </span>
                    )}
                </div>
                <div className="lg:flex hidden">
                    <button
                        className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-full hover:bg-yellow-300 hover:animate-pulse"
                        onClick={onClick}
                    >
                        <FaPlus />
                    </button>
                </div>
            </div>
            <div className="flex lg:hidden absolute bottom-1 z-[200] right-1">
                <button
                    className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-full hover:bg-yellow-300"
                    onClick={onClick}
                >
                    <FaPlus />
                </button>
            </div>
        </div>
    );
};

export default ItemCard;
