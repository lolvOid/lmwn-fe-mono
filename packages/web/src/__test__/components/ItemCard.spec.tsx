import { render, screen, fireEvent } from '@testing-library/react';
import ItemCard from '@/components/card/ItemCard';
import renderer from 'react-test-renderer';

describe('ItemCard', () => {
    const mockProps = {
        name: 'Cake',
        imageSource: 'sample-image.jpg',
        fullPrice: 100,
        discountedPercent: '10',
        totalInStock: 5,
        onClick: jest.fn(),
    };

    it('Should render and match the snapshot with mockProps', () => {
        const itemCard = renderer.create(<ItemCard {...mockProps} />).toJSON();
        expect(itemCard).toMatchSnapshot();
    });

    it('Render properly with provided props', () => {
        render(<ItemCard {...mockProps} />);
        expect(screen.getByTestId('item-card-name')).toHaveTextContent('Cake');
        expect(screen.getByAltText('Cake')).toBeInTheDocument();
        expect(screen.getByTestId('item-card-discounted-price')).toHaveTextContent('฿90');
        expect(screen.getByTestId('item-card-price')).toHaveTextContent('฿100');
    });

    it('Should call onClick when clicked', () => {
        render(<ItemCard {...mockProps} />);
        fireEvent.click(screen.getByText('Cake'));
        expect(mockProps.onClick).toHaveBeenCalled();
    });
});
