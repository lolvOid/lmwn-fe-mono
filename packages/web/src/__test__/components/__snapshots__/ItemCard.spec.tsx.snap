// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard Should render and match the snapshot with mockProps 1`] = `
<div
  className="relative flex lg:flex-row flex-col lg:justify-between border-solid border-gray-200 border-[0.4px] p-2 rounded-md opacity-100 cursor-pointer hover:bg-gray-100"
  data-testid="item-card"
  onClick={[Function]}
>
  <div
    className="lg:w-32 w-full max-w-40 h-32 flex-shrink-0 justify-center items-center"
  >
    <img
      alt="Cake"
      className="w-full h-full object-cover object-center"
      src="sample-image.jpg"
    />
  </div>
  <div
    className="px-2 py-2 flex lg:flex-col w-full flex-grow lg:items-end items-start"
  >
    <div
      className="w-full flex flex-col text-sm flex-grow gap-1"
    >
      <span
        className="text-xs text-wrap"
        data-testid="item-card-name"
      >
        Cake
      </span>
      <span
        className="text-sm primary-text text-wrap"
        data-testid="item-card-discounted-price"
      >
        ฿90
      </span>
      <span
        className="text-xs text-wrap text-gray-600 line-through"
        data-testid="item-card-price"
      >
        ฿100
      </span>
    </div>
    <div
      className="lg:flex hidden"
    >
      <button
        className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-full hover:bg-yellow-300 hover:animate-pulse"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="flex lg:hidden absolute bottom-1 z-[200] right-1"
  >
    <button
      className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-full hover:bg-yellow-300"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        />
      </svg>
    </button>
  </div>
</div>
`;
